Markmirror: a markdown editor

pabzm opened the pull request:
This adds another email composing editor to Roundcubemail, which converts its content to HTML before sending.

It provides syntax highlighting and a toolbar (including a preview button) to help writing markdown. Drafts are saved as-is, and are automatically re-opened in this editor on re-editing. On sending the written markdown text gets converted into HTML.

(Roundcubemail automatically produces a multipart/alternative email from the given HTML, including a text/plain part consisting of re-converted HTML-to-text. Using our original text as text/plain part would be preferable but is left for a future version of this plugin.)

Technically it is a codemirror v6 configured for markdown with a custom toolbar and a preview function (and some glue code). Codemirror v6 isn't provided as a zipfile download anymore, but only via NPM, so this is using NPM, too.

It is built as a plugin to be easily ignorable. I do think that with a little bit of work prosemirror could be a good replacement for TinyMCE at some point, though (it would make us less dependent on an increasingly commercially driven 3rd party project). For that task this code might be a helpful inspiration.

This PR includes a few changes to core code, which were unavoidable but don't change any behaviour as far as I can see, and might even be helpful for other code authors (e.g. being able to hook into the messageform submission; and allowing more modern Javascript).

Closes #6276 

-----------------


<img width="569" height="348" alt="Screenshot 2025-09-04 at 18 39 43" src="https://github.com/user-attachments/assets/dca9d347-18bd-4f15-8ce7-004188343c69" />


----------------

<img width="1057" height="706" alt="Screenshot 2025-09-04 at 18 41 20" src="https://github.com/user-attachments/assets/03d76c07-1517-4620-a31a-710905f2290b" />

----------------

<img width="1058" height="703" alt="Screenshot 2025-09-04 at 18 41 31" src="https://github.com/user-attachments/assets/b34b6cd8-d504-4e70-bb12-451141b0a1f8" />

pabzm committed:
New plugin "markdown_editor": compose in markdown, send as HTML

This adds a markdown editor that sends HTML to the server.

It uses codemirror and some custom code to show a syntax highlighted
textarea and some buttons to help editing
(including a preview).

Drafts get marked via an internal email header that causes the markdown
editor to automatically start if a message composition is
continued that was started using the markdown editor.

pabzm committed:
Allow event handlers to hook into messageform submission

submit() doesn't emit an event prior to sending the data but
requestSubmit() does.

pabzm committed:
Allow event handlers to act after a signature was inserted

pabzm committed:
Allow additional attributes for included scripts

pabzm committed:
Make eslint check ECMAscript 2022

This allows using public and private class fields, which are supported in all major browsers since years, (but just made
it into ECMAScript 2022).
See
<https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields#browser_compatibility>
and <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_elements#browser_compatibility>.

alecpl commented:
1. I've found it in logs "PHP Warning:  Undefined property: stdClass::$headers in /home/alec/repos/roundcubemail/plugins/markmirror/markmirror.php on line 26".

2. `make plugins-build` ends with:
```
The file undefined was skipped because it was not found and the import was marked optional.
Skipped data-uri embedding of ../images/corner-handle.svg because file not found
Skipped data-uri embedding of ../images/download.svg because file not found
lessc: wrote /home/alec/repos/roundcubemail/plugins/markmirror/markmirror.min.css
The file undefined was skipped because it was not found and the import was marked optional.
Skipped data-uri embedding of ../images/corner-handle.svg because file not found
Skipped data-uri embedding of ../images/download.svg because file not found
lessc: wrote /home/alec/repos/roundcubemail/plugins/markmirror/iframe.min.css
```
Aside the issues here, I'd like the css, less, images in plugins/markmirror/skins/elastic

3. `composer.json` file is missing.

4. I do have `markmirror.min.js` file, but `https://roundcube.test/static.php/plugins/markmirror/markmirror.js` is being requested that does not exist.

So, I wasn't able to test this yet.

alecpl commented:
Ok, just renamed the js file and was able to test it:

5. Underline feature is broken. It creates `<u>text</u>`, which in preview looks plain text with tags. Same if you close the markdown editor going to plain text.
6. The blockquote needs probably some styles to look better (in preview).
7. Spellchecking is not supported in markmirror.
8. Bug when switching into markmirror editor from the plain text editor with Spellcheck enabled.
9. Undo/Redo buttons have broken labels.
10. Including images does not work properly. Creates one long line that makes the right side of the toolbar unreachable.
11. The editor area is not auto-resizing as in plain text mode.
12. The editor toolbar is not usable on mobile.

pabzm committed:
Fix seeings screenshots on host for containerized browser tests with run-in-copy.sh

pabzm commented:
1. is fixed, it occur when editing a fresh message

alecpl commented:
6. We add some inline styles: https://github.com/roundcube/roundcubemail/blob/c2b785f78b958551b07adec1073df01a69b7b1f1/program/actions/mail/send.php#L122
10. I would remove the image feature for now.
12. Yes, the buttons in mobile are too close.
13. If you have a long content, then click Preview, the textarea resizes to a smaller height.
14. I'd like the same font-size/line-height in the markdown and plain text modes. So, there's no "text jumps" when you switch between these modes.
15. Inserting a response (the Responses menu) puts them at the end of the whole text instead at the cursor position.
16. Fontawesome has nicer undo/redo icons.
17. The textarea is cut on screen width between 481px and ~640px.

pabzm commented:
6. Thank you for the pointer, the iframe now shows the same styles.
7. Agreed & removed
8. I changed the toolbar so the buttons wrap around into a second row if the screen gets too narrow. That styling could still be improved but I think for a first version this solution is good enough.
9. Fixed.
10. Good idea, I brought the font styling in line, thanks! 
11. Hm, I'll have to think about how to fix this.
12. Not in the version of Font Awesome we're shipping currently (5.9.0) AFAICT. If you're sure: could you let me know how to display them? `@fa-var-undo` (as defined in fontawesome.less) doesn't work.
13. I don't understand what you mean and didn't spot problematic behaviour myself, could you explain please?

pabzm commented on program/js/app.js:
```diff
@@ -5108,7 +5108,7 @@ function rcube_webmail() {
         }, this.env.request_timeout * 1000);
 
-        form.submit();
+        form.requestSubmit();
```
That's required so [other code](https://github.com/roundcube/roundcubemail/blob/markmirror/plugins/markmirror/javascript/index.js#L216) can hook into the form submission. `submit()` doesn't emit an event.

pabzm commented on Makefile:
```diff
@@ -112,3 +112,17 @@ downloads-json-data:
 
 release: edit-changelog git-tag all sign verify downloads-json-data git-tag-push
+
+plugins-build: plugin-markmirror-build
```
We also could commit the built JS and CSS to the repo, then the building step and rollup would only be required for development.

alecpl commented:
>     12. Not in the version of Font Awesome we're shipping currently (5.9.0) AFAICT. If you're sure: could you let me know how to display them? `@fa-var-undo` (as defined in fontawesome.less) doesn't work.

Both @fa-var-undo and @fa-var-redo work for me.

alecpl commented:
>     13. I don't understand what you mean and didn't spot problematic behaviour myself, could you explain please?

Seems the menu changes fixed it.

pabzm commented on plugins/markmirror/markmirror.php:
```diff
@@ -0,0 +1,57 @@
+        if (isset($args['fetch_headers'])) {
+            $args['fetch_headers'] .= ' X-EDITED-BY-MARKMIRROR';
+        } else {
+            $args['fetch_headers'] = 'X-EDITED-BY-MARKMIRROR';
```
Alright, then I'll put it aside for now. But I agree it's would be a nice idea for a future change!

alecpl reviewed (approved):
Please, update install/upgrade procedure with `make plugins-build` step.

We should probably consider simplifying/merging all javascript/css related steps into a single Makefile target. And use that in the docs.

pabzm commented:
@alecpl Please let me know what you think.

pabzm committed:
Combine some install-steps in one make target
